src/rewards/RewardsManager.sol

```diff
@@ -0,0 +1,91 @@
+// SPDX-License-Identifier: MIT
+pragma solidity 0.8.17;
+
+interface IRewardsManager {
+    event RewardsAdded(address indexed recipient1, uint256 amount1, address indexed recipient2, uint256 amount2, address indexed recipient3, uint256 amount3);
+    event RewardsAdded(
+        address indexed recipient1,
+        uint256 amount1,
+        address indexed recipient2,
+        uint256 amount2,
+        address indexed recipient3,
+        uint256 amount3,
+        address recipient4,
+        uint256 amount4
+    );
+
+    error INVALID_AMOUNT();
+    error FAILED_WITHDRAW();
+
+    function addReward(address recipient1, uint256 amount1, address recipient2, uint256 amount2, address recipient3, uint256 amount3) external payable;
+
+    function addReward(
+        address recipient1,
+        uint256 amount1,
+        address recipient2,
+        uint256 amount2,
+        address recipient3,
+        uint256 amount3,
+        address recipient4,
+        uint256 amount4
+    ) external payable;
+
+    function withdrawReward() external;
+}
+
+/** TODO
+- variants of addReward params and events
+- withdraw w/ recipient
+- withdraw w/ sig?
+ */
+contract RewardsManager is IRewardsManager {
+    mapping(address => uint256) public ethBalance;
+
+    function addReward(address recipient1, uint256 amount1, address recipient2, uint256 amount2, address recipient3, uint256 amount3) external payable {
+        if (msg.value != (amount1 + amount2 + amount3)) {
+            revert INVALID_AMOUNT();
+        }
+
+        ethBalance[recipient1] += amount1;
```

Comment:
there are now three storage writes and adds to track rewards, even in the case there is no finder or lister.   I'd have an overload that just writes to a single ethBalance, for the most common case when there is no finder or lister, and call that when there is no finder or lister.